### ISSUE TITLE
TMP Stopped Working On Waterfox G4.1.3

### ISSUE THREAD

**Issue by thunderstone135** [onemen/TabMixPlus#150]
Hi, I just want to let you know that TMP 1.0.0-pre.8 stopped working on G4.1.3.
I was surprised when I noticed that TMP features went missing so I checked my Addon Page.
As it turns out, TMP got disabled. I tried updating/reinstalling, but it's telling me the addon was corrupt.
Waterfox G4.1.2 is not affected, so I went back to that version. In any event, I also raised the issue **[here](https://github.com/WaterfoxCo/Waterfox/issues/2725)**.

**Comment by onemen** [comment]
I see this error in the console

```
TypeError: "freeze" is read-only[BootstrapLoader.jsm:79:1](resource:///modules/BootstrapLoader.jsm)
```

It looks like Waterfox BootstrapLoader.jsm uses old code 
```
const Object = Cu.getGlobalForObject(Cu).Object;
const { freeze } = Object;
Object.freeze = obj => {
  if (
    Components.stack.caller.filename !=
      "resource://gre/modules/AddonManager.jsm" ||
    !obj.OPTIONS_TYPE_TAB
  )
    return freeze(obj);

  obj.OPTIONS_TYPE_DIALOG = 1;
  Object.freeze = freeze;
  return freeze(obj);
}
```

**Comment by thunderstone135** [comment]
> Waterfox use a integrated legacy extension loader of their own. Sorry but no one but themself can take care of it.
> 
> However It may be possible to install UserChromeJS for loader if the inner one is completely stop working on TMP but it still likely to break other legacy extension.
I tried installing UserChromeJS yesterday, but it can't seem to load the script below from config.js:
```
// signing bypass by onemen
const Constants = ChromeUtils.import('resource://gre/modules/AppConstants.jsm');
const temp = Object.assign({}, Constants.AppConstants);
temp.MOZ_REQUIRE_SIGNING = false
Constants.AppConstants = Object.freeze(temp);

try {
  let cmanifest = Cc['@mozilla.org/file/directory_service;1'].getService(Ci.nsIProperties).get('UChrm', Ci.nsIFile);
  cmanifest.append('utils');
  cmanifest.append('chrome.manifest');
  Components.manager.QueryInterface(Ci.nsIComponentRegistrar).autoRegister(cmanifest);

  Cu.import('chrome://userchromejs/content/BootstrapLoader.jsm');
} catch (ex) {};

try {
  Cu.import('chrome://userchromejs/content/userChrome.jsm');
} catch (ex) {};
```
Waterfox just complains about not being able to read the prefs.
If I remove the script posted above, it's able to load.

**Comment by 117649** [comment]
> Waterfox just complains about not being able to read the prefs. If I remove the script posted above, it's able to load.
> 
> AFAICT, @xiaoxiaoflood mentioned that he only support Firefox Dev and not Waterfox. I can't blame him since there's most likely a huge difference in coding because Waterfox is based on ESR.

Do you get it working after all?

> It looks like Waterfox BootstrapLoader.jsm uses old code

In fact you can edit 'omni.ja' by unzip it with 7-zip in order to edit "resource:///modules/BootstrapLoader.jsm"

**Comment by thunderstone135** [comment]
> Do you get it working after all?

Nope. Without the signing bypass script by onemen UserChromeJS simply doesn't work.
In the meantime, I reverted back to G4.1.2 to make TMP working again.
I'll wait until the Waterfox Team fixes the issue before I update.

> In fact you can edit 'omni.ja' by unzip it with 7-zip in order to edit "resource:///modules/BootstrapLoader.jsm"

I compared the BootstrapLoader.jsm of G4.1.2 and G4.1.3 just now, and they're completely identical 🤔 

![Imgur](https://imgur.com/TWx2mlb.png)

It's just a hunch, but I think it has something to do with security fixes that were merged from Firefox to Waterfox.
